app/models/status.rb

```diff
@@ -101,11 +101,15 @@ class Status < ApplicationRecord
   scope :including_silenced_accounts, -> { left_outer_joins(:account).where.not(accounts: { silenced_at: nil }) }
   scope :not_excluded_by_account, ->(account) { where.not(account_id: account.excluded_from_timeline_account_ids) }
   scope :not_domain_blocked_by_account, ->(account) { account.excluded_from_timeline_domains.blank? ? left_outer_joins(:account) : left_outer_joins(:account).where('accounts.domain IS NULL OR accounts.domain NOT IN (?)', account.excluded_from_timeline_domains) }
-  scope :tagged_with_all, lambda { |tag_ids|
-    Array(tag_ids).map(&:to_i).reduce(self) do |result, id|
-      result.joins("INNER JOIN statuses_tags t#{id} ON t#{id}.status_id = statuses.id AND t#{id}.tag_id = #{id}")
-    end
+  scope :tagged_with_all, ->(tag_ids) { tag_ids.present? ? where(id: tagged_with_all_ids(tag_ids)) : nil }
+  scope :tagged_with_all_ids, lambda { |tag_ids|
+    select('statuses_tags.status_id')
+      .from('statuses_tags')
+      .where(statuses_tags: { tag_id: tag_ids })
+      .group('statuses_tags.status_id')
+      .having('COUNT(DISTINCT statuses_tags.tag_id) = ?', tag_ids.size)
   }
+
```

Comment:
While the query seems logically correct, I believe this to be prohibitively expensive, as it will have to go through *all* of the tags usage of any of those tags for each single page, without being able to make meaningful use of `LIMIT`.

A quick check on my single-user's database against two somewhat popular tags that are seldom used together returns one element within the second on the old query, and doesn't return in minutes with the new one.